bugfix: core: Improve consistency while toggling from topic to stream.

Abhirup-99 opened the pull request:
Fixes #934.

Abhirup-99 commented:
@zulipbot add "PR needs review"

Ezio-Sarthak reviewed (commented):
Thanks for taking this up! The toggle behavior seems OK now :+1: 

As you rightly pointed out about the gap in 2 sets of indexed messages, I believe fixing that should fix this too?

I also found another fix locally for the stream toggle bug. Discussion [link](https://chat.zulip.org/#narrow/stream/206-zulip-terminal/topic/Stream.20toggle.20bugfix.20(.23T934)) :)

Abhirup-99 committed:
bugfix: core: Improve consistency while toggling from topic to stream.

The commit brings consistency while toggling from topic
to streams. The original behaviour was to move to the last
message in the stream, while toggling out. This changes it to
zoom out to the context of the message selected in the stream.
Tests added.
Fixes #934.

neiljp commented:
@Abhirup-99 I think we should avoid blowing away the message cache if we can - though I appreciate that this does look to be a solution - as we could end up fetching messages multiple times if nothing else. If we did take this approach, we would likely want to isolate the changes in the model in any case, as the change digs into the model from the controller; code like that still exists in ZT, but it's a lot easier to reason about otherwise.

See the topic @Ezio-Sarthak linked for other discussion.
